tmc-plugin/src/fi/helsinki/cs/tmc/exerciseSubmitter/ExerciseSubmitter.java

```diff
@@ -77,53 +75,48 @@ private void submitProject(final TmcProjectInfo project) {
         projectMediator.saveAllFiles();
 
         // Oh what a mess :/
-
         final SubmissionResultWaitingDialog dialog = SubmissionResultWaitingDialog.createAndShow();
 
-        final BgTaskListener<ServerAccess.SubmissionResponse> submissionListener = new BgTaskListener<ServerAccess.SubmissionResponse>() {
-            @Override
-            public void bgTaskReady(ServerAccess.SubmissionResponse response) {
-                final SubmissionResultWaiter waitingTask = new SubmissionResultWaiter(response.submissionUrl.toString(), dialog);
-                dialog.setTask(waitingTask);
+        CancellableCallable<SubmissionResult> submitAndPollTask = new CancellableCallable<SubmissionResult>() {
 
-                BgTask.start("Waiting for results from server.", waitingTask, new BgTaskListener<SubmissionResult>() {
+            ListenableFuture<SubmissionResult> lf;
 
-                    @Override
-                    public void bgTaskReady(SubmissionResult result) {
+            @Override
+            public SubmissionResult call() throws Exception {
+                lf = TmcCoreSingleton.getInstance().submit(project.getProjectDirAsPath());
+                return lf.get();
+            }
 
-                        dialog.close();
+            @Override
+            public boolean cancel() {
+                return lf.cancel(true);
```

Comment:
Possible NPE if cancelled is called before task is started (due to concurrent task limit or something)